src/Backend/SgateShopgatePlugin/Controllers/Frontend/shopgate.php

```diff
@@ -288,4 +377,988 @@ public function pluginAction()
+                'Please select an option to place the required product in the cart', true);
+        }
+
+        $quantity = max(1, (int) $quantity);
+        $inStock = $this->getAvailableStock($orderNumber);
+
+        $inStock['quantity'] = $quantity;
+
+        if (empty($inStock['articleID'])) {
+            return Shopware()->Snippets()->getNamespace('frontend')->get('CheckoutArticleNotFound',
+                'Product could not be found.', true);
+        }
+        if (!empty($inStock['laststock']) || !empty(Shopware()->Config()->InstockInfo)) {
+            if ((int)$inStock['instock'] <= 0 && !empty($inStock['laststock'])) {
+                return Shopware()->Snippets()->getNamespace('frontend')->get('CheckoutArticleNoStock',
+                    'Unfortunately we can not deliver the desired product in sufficient quantity', true);
+            } elseif ((int)$inStock['instock'] < (int)$inStock['quantity']) {
+                $result = 'Unfortunately we can not deliver the desired product in sufficient quantity. (#0 of #1 in stock).';
+                $result = Shopware()->Snippets()->getNamespace('frontend')->get('CheckoutArticleLessStock', $result,
+                    true);
+
+                return str_replace(['#0', '#1'], [$inStock['instock'], $inStock['quantity']], $result);
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Get current stock from a certain product defined by $ordernumber
+     * Support for multidimensional variants
+     *
+     * @param unknown_type $ordernumber
+     *
+     * @return array with article id / current basket quantity / instock / laststock
+     */
+    private function getAvailableStock($ordernumber)
+    {
+        $sql = '
+            SELECT
+                a.id as articleID,
+                ob.quantity,
+                IF(ad.instock < 0, 0, ad.instock) as instock,
+                a.laststock,
+                ad.ordernumber as ordernumber
+            FROM s_articles a
+            LEFT JOIN s_articles_details ad
+            ON ad.ordernumber=?
+            LEFT JOIN s_order_basket ob
+            ON ob.sessionID=?
+            AND ob.ordernumber=ad.ordernumber
+            AND ob.modus=0
+            WHERE a.id=ad.articleID
+        ';
+        $row = Shopware()->Db()->fetchRow($sql, [
+            $ordernumber,
+            Shopware()->Session()->get('sessionId'),
+        ]);
+
+        return $row;
+    }
+
+    /**
+     * Decodes a JWT string into a PHP object.
+     *
+     * @param string        $jwt            The JWT
+     * @param string|array  $key            The key, or map of keys.
+     *                                      If the algorithm used is asymmetric, this is the public key
+     *
+     * @return object The JWT's payload as a PHP object
+     *
+     * @throws UnexpectedValueException     Provided JWT was invalid
+     *
+     * @uses urlsafeB64Decode
+     */
+    public static function jwtDecode($jwt, $key)
```

Comment:
Any reason why is in shopgate.php file? Maybe use a composer module or outsource to a different file instead of continuing the sphagetti code.

Reply:
I can't use composer because a shopware plugin can't install composer dependencies. I first used composer but in an earlier pull request I had to remove this dependency. Well I can outsource it to an extra file.